File: ports/python/arkose_solver/devices.py
"""Per-solve device generation from the embedded weighted population.

devices.json is the same file the Go implementation embeds: 234 GPUs,
105 screens, 13 core counts, 4 memory sizes, 62 connection profiles.
"""

import json
import os
import random

_HERE = os.path.dirname(os.path.abspath(__file__))
_POOL = None

CHROME_COLOR_DEPTH = 24
INNER_WIDTH_INSET = 16
INNER_HEIGHT_INSET = 95


def _pool():
    global _POOL
    if _POOL is None:
        with open(os.path.join(_HERE, "devices.json"), encoding="utf-8") as fh:
            _POOL = json.load(fh)
        for name in ("gpus", "screens", "cores", "memory", "network"):
            if not _POOL.get(name):
                raise RuntimeError("devices.json is missing pool %r" % name)
    return _POOL


def _pick(entries):
    total = sum(e["weight"] for e in entries if e.get("weight", 0) > 0)
    if total <= 0:
        return entries[0]
    r = random.randrange(total)
    for e in entries:
        w = e.get("weight", 0)
        if w <= 0:
            continue
        r -= w
        if r < 0:
            return e
    return entries[0]


def js_heap_size_limit(device_memory_gb):
    """enhanced_fp["basfas"] carries performance.memory.jsHeapSizeLimit."""
    if device_memory_gb >= 8:
        return 4395630592
    if device_memory_gb >= 4:
        return 4294705152
    return 2197815296


def _window_on(screen):
    avail_w = screen.get("availWidth") or screen["width"]
    avail_h = screen.get("availHeight") or screen["height"]
    if random.randrange(100) < 35:
        return avail_w, avail_h
    w = avail_w * random.randint(60, 100) // 100
    h = avail_h * random.randint(60, 100) // 100
    w = max(w, INNER_WIDTH_INSET + 200)
    h = max(h, INNER_HEIGHT_INSET + 200)
    return w, h


def generate_device():
    """A fresh, internally coherent synthetic machine."""
    pool = _pool()
    gpu = _pick(pool["gpus"])
    screen = _pick(pool["screens"])
    net = _pick(pool["network"])
    outer_w, outer_h = _window_on(screen)

    return {
        "gpu_vendor": gpu["vendor"],
        "gpu_model": gpu["model"],
        "gl_vendor": gpu["glVendor"],
        "gl_renderer": gpu["glRenderer"],
        "screen_width": screen["width"],
        "screen_height": screen["height"],
        "avail_width": screen.get("availWidth") or screen["width"],
        "avail_height": screen.get("availHeight") or screen["height"],
        "color_depth": CHROME_COLOR_DEPTH,
        "device_pixel_ratio": screen.get("dpr", 1),
        "outer_width": outer_w,
        "outer_height": outer_h,
        "inner_width": outer_w - INNER_WIDTH_INSET,
        "inner_height": outer_h - INNER_HEIGHT_INSET,
        "device_memory": _pick(pool["memory"])["value"],
        "hardware_concurrency": _pick(pool["cores"])["value"],
        "downlink": net["downlink"],
        "rtt": net["rtt"],
        "save_data": net.get("saveData", False),
    }


def pool_sizes():
    p = _pool()
    return {k: len(p[k]) for k in ("gpus", "screens", "cores", "memory", "network")}

File: ports/node/src/devices.js
'use strict';

// Per-solve device generation from the shared weighted population.
// devices.json is the same file the Go implementation embeds.

const fs = require('fs');
const path = require('path');

const CHROME_COLOR_DEPTH = 24;
const INNER_WIDTH_INSET = 16;
const INNER_HEIGHT_INSET = 95;

let _pool = null;
function pool() {
  if (_pool === null) {
    _pool = JSON.parse(fs.readFileSync(path.join(__dirname, 'devices.json'), 'utf8'));
    for (const name of ['gpus', 'screens', 'cores', 'memory', 'network']) {
      if (!_pool[name] || !_pool[name].length) {
        throw new Error(`devices.json is missing pool ${name}`);
      }
    }
  }
  return _pool;
}

function pick(entries) {
  let total = 0;
  for (const e of entries) if (e.weight > 0) total += e.weight;
  if (total <= 0) return entries[0];
  let r = Math.floor(Math.random() * total);
  for (const e of entries) {
    if (!(e.weight > 0)) continue;
    r -= e.weight;
    if (r < 0) return e;
  }
  return entries[0];
}

// enhanced_fp["basfas"] carries performance.memory.jsHeapSizeLimit.
function jsHeapSizeLimit(deviceMemoryGB) {
  if (deviceMemoryGB >= 8) return 4395630592;
  if (deviceMemoryGB >= 4) return 4294705152;
  return 2197815296;
}

function windowOn(screen) {
  const availW = screen.availWidth || screen.width;
  const availH = screen.availHeight || screen.height;
  if (Math.floor(Math.random() * 100) < 35) return [availW, availH];
  const pct = () => 60 + Math.floor(Math.random() * 41);
  let w = Math.floor((availW * pct()) / 100);
  let h = Math.floor((availH * pct()) / 100);
  w = Math.max(w, INNER_WIDTH_INSET + 200);
  h = Math.max(h, INNER_HEIGHT_INSET + 200);
  return [w, h];
}

function generateDevice() {
  const p = pool();
  const gpu = pick(p.gpus);
  const screen = pick(p.screens);
  const net = pick(p.network);
  const [outerW, outerH] = windowOn(screen);

  return {
    gpuVendor: gpu.vendor,
    gpuModel: gpu.model,
    glVendor: gpu.glVendor,
    glRenderer: gpu.glRenderer,
    screenWidth: screen.width,
    screenHeight: screen.height,
    availWidth: screen.availWidth || screen.width,
    availHeight: screen.availHeight || screen.height,
    colorDepth: CHROME_COLOR_DEPTH,
    devicePixelRatio: screen.dpr === undefined ? 1 : screen.dpr,
    outerWidth: outerW,
    outerHeight: outerH,
    innerWidth: outerW - INNER_WIDTH_INSET,
    innerHeight: outerH - INNER_HEIGHT_INSET,
    deviceMemory: pick(p.memory).value,
    hardwareConcurrency: pick(p.cores).value,
    downlink: net.downlink,
    rtt: net.rtt,
    saveData: !!net.saveData,
  };
}

function poolSizes() {
  const p = pool();
  const out = {};
  for (const k of ['gpus', 'screens', 'cores', 'memory', 'network']) out[k] = p[k].length;
  return out;
}

module.exports = {
  CHROME_COLOR_DEPTH, INNER_WIDTH_INSET, INNER_HEIGHT_INSET,
  generateDevice, jsHeapSizeLimit, poolSizes,
};
